Quarkus native heap dump on OOME blog post

galderz opened the pull request:
Blog post on how to use heap dump on OOME option with Quarkus native which touches a few extra related topics:

* How to configure epsilon GC.
* How to inspect heap dumps with JOL.
* Compare heap dumps with JVM mode.

galderz commented:
Can we add @roberttoyonaga as reviewer?

zakkak commented on _posts/2024-04-15-heapdump-oome-native.adoc:
```diff
@@ -0,0 +1,258 @@
+Once the Quarkus application is running with Epsilon GC,
+apply some load and within a short space of time it will run out of memory and produce a heap dump.
+
+Let’s start by producing a heap dump with a Quarkus native executable.
+A good starting point would be a Quarkus application that simply responds to an HTTP endpoint request, e.g. the
+https://github.com/quarkusio/quarkus-quickstarts/tree/main/getting-started-reactive[getting-started-reactive] Quarkus quickstart.
+Once checked out,
+build a Quarkus native executable with GraalVM for JDK 21 configuring Epsilon GC.
+For Quarkus native mode, selecting the GC needs to be done at build time:
+
+[source,bash]
+----
+$ ./mvnw package -DskipTests -Dnative -Dquarkus.native.additional-build-args=--gc=epsilon
+----
+
```
I think this can be shortened a bit:

```suggestion
Let’s do this using a Quarkus application that simply responds to an HTTP endpoint request as a starting point, e.g. the
https://github.com/quarkusio/quarkus-quickstarts/tree/main/getting-started-reactive[getting-started-reactive] Quarkus quickstart.
Once checked out,
we build a Quarkus native executable with GraalVM for JDK 21 configuring it to use Epsilon GC:

[source,bash]
----
$ ./mvnw package -DskipTests -Dnative -Dquarkus.native.additional-build-args=--gc=epsilon
----

Note: The GC selection needs to be done at build time for Quarkus native applications.

```

zakkak commented on _posts/2024-04-15-heapdump-oome-native.adoc:
```diff
@@ -0,0 +1,258 @@
+2024-04-12 09:58:00,855 INFO  [io.quarkus] (main) Installed features: [cdi, rest, smallrye-context-propagation, vertx]
+----
+
+Next, apply some load to the HTTP endpoint exposed by the Quarkus application, for example using `wrk`:
```
Perhaps a bash loop with a curl command would be better as most people won't have `wrk` installed. On the other hand that's a nice way to become aware of `wrk`.

galderz replied:
Good point, `wrk` is an overkill for this. Just used to quickly firing it up :)

zakkak commented on _posts/2024-04-15-heapdump-oome-native.adoc:
```diff
@@ -0,0 +1,258 @@
+
+Let’s start by producing a heap dump with a Quarkus native executable.
+A good starting point would be a Quarkus application that simply responds to an HTTP endpoint request, e.g. the
+https://github.com/quarkusio/quarkus-quickstarts/tree/main/getting-started-reactive[getting-started-reactive] Quarkus quickstart.
```
Perhaps it's simpler to use https://code.quarkus.io/

E.g. I can follow this post with:

```
wget "https://code.quarkus.io/d?S=io.quarkus.platform%3A3.8&cn=code.quarkus.io" -O code.zip
unzip code.zip
cd code-with-quarkus
```

Note I explicitly set the version to Quarkus 3.8 to hopefully keep the instructions working for some time.

zakkak commented on _posts/2024-04-15-heapdump-oome-native.adoc:
```diff
@@ -0,0 +1,258 @@
+
+[source,bash]
+----
+$ java -XX:+HeapDumpOnOutOfMemoryError -XX:+ExitOnOutOfMemoryError -XX:+UnlockExperimentalVMOptions -XX:+UseEpsilonGC -Xmx64m -jar quarkus-run.jar
```
```suggestion
$ mvnw package -DskipTests
$ java -XX:+HeapDumpOnOutOfMemoryError -XX:+ExitOnOutOfMemoryError -XX:+UnlockExperimentalVMOptions -XX:+UseEpsilonGC -Xmx64m -jar quarkus-run.jar
```

zakkak reviewed (approved):
LGTM, thank @galderz 

I added some suggestions, for your consideration, that IMHO might improve it a bit.

roberttoyonaga commented on _posts/2024-04-15-heapdump-oome-native.adoc:
```diff
@@ -0,0 +1,258 @@
+
+Starting with GraalVM for JDK 21,
+native executables can run with the `-XX:+HeapDumpOnOutOfMemoryError` option to generate a heap dump when a `java.lang.OutOfMemoryError` is thrown.
+In this blog post we will explore how to use the flag, we will inspect what a native image heap dump looks like and how it compares with one produced by HotSpot.
```
>we will inspect what a native image heap dump looks like

Maybe there's a chance reader's  get confused by the wording here and think dumping the  Image Heap and "java heap" are different things (If they are aware that the image heap exists).

Maybe it would be good to specify that we're dumping both the read-only Image Heap as well as the native image's "Java heap". Whereas in Java mode we're just dumping the Java heap.

roberttoyonaga replied:
On the flip side, it could also be an opportunity to briefly inform reader's who aren't aware that there is a special segmentation to the heap in native mode.

galderz replied:
Both points sound good. I have not explored the differences in detail between the read-only and the read/write parts, but if you have feel free to propose a paragraph and add it as suggestion and we can just add it.

roberttoyonaga replied:
ok I've added a suggestion below!

roberttoyonaga reviewed (approved):
Looks good to me! I think this is clear and to-the-point.

roberttoyonaga commented on _posts/2024-04-15-heapdump-oome-native.adoc:
```diff
@@ -0,0 +1,259 @@
+
+Starting with GraalVM for JDK 21,
+native executables can run with the `-XX:+HeapDumpOnOutOfMemoryError` option to generate a heap dump when a `java.lang.OutOfMemoryError` is thrown.
+In this blog post we will explore how to use the flag, we will inspect what a native image heap dump looks like and how it compares with one produced by HotSpot.
```
```suggestion
In this blog post we will explore how to use the flag, we will inspect what a Native Image heap dump looks like and how it compares with one produced by HotSpot.

The heap of Native Image executables has both read-only and read-write segments. The read-only part is referred to as the "image heap" and contains the data pre-initialized during the image build to help speed up start-up time. The read-write part is where allocations at runtime are made. Generating a Native Image heap dump will contain content from both the "image heap" and the read-write heap.   
```

galderz commented:
@zakkak I've addressed all your comments. @roberttoyonaga I've integrated your suggestion and made some small changes to it, let me know if you're happy with it.

zakkak commented on _posts/2024-04-15-heapdump-oome-native.adoc:
```diff
@@ -0,0 +1,265 @@
+Let’s do this using a Quarkus application that simply responds to an HTTP endpoint request as a starting point.
+The sample application can be downloaded from
+https://code.quarkus.io[code.quarkus.io] using a browser
+or you via the command line:
```
```suggestion
or via the command line:
```

zakkak commented on _posts/2024-04-15-heapdump-oome-native.adoc:
```diff
@@ -0,0 +1,265 @@
+In this blog post we will explore how to use the flag,
+we will inspect what a GraalVM Native Image heap dump looks like and how it compares with one produced by HotSpot.
+
+The heap of GraalVM Native Image executables has both read-only and read-write segments.
+The read-only part is referred to as the "image heap" and contains the data pre-initialized during the image build to help speed up start-up time.
+The read-write part is where allocations at runtime are made.
+Therefore, generating heap dump at runtime will contain content from both the "image heap" and the read-write heap.
```
Perhaps add this as a "note"?

```suggestion
Note: The heap of GraalVM Native Image executables has both read-only and read-write segments.
The read-only part is referred to as the "image heap" and contains the data pre-initialized during the image build to help speed up start-up time.
The read-write part is where allocations at runtime are made.
Therefore, generating heap dump at runtime will contain content from both the "image heap" and the read-write heap.
```

zakkak reviewed (changes requested):
Thanks for addressing my comments Galder. 

There seems to be a new typo introduced.

roberttoyonaga commented:
> I've integrated your suggestion and made some small changes to it, let me know if you're happy with it.

Looks good to me!

galderz commented:
Squashed and pushed commits.

gsmet commented on _posts/2024-04-15-heapdump-oome-native.adoc:
```diff
@@ -0,0 +1,265 @@
+layout: post
+title: 'Obtaining heap dump on OutOfMemoryError with Quarkus native'
+date: 2024-04-15
+tags: native heapdump oome epsilongc jol
```
I don't think it's a good idea to add tags that are so precise. We already have too many tags. Let's only have `native`?

gsmet commented on _posts/2024-04-15-heapdump-oome-native.adoc:
```diff
@@ -0,0 +1,265 @@
+Note: The heap of GraalVM Native Image executables has both read-only and read-write segments.
+The read-only part is referred to as the "image heap" and contains the data pre-initialized during the image build to help speed up start-up time.
+The read-write part is where allocations at runtime are made.
+Therefore, generating heap dump at runtime will contain content from both the "image heap" and the read-write heap.
```
```suggestion
Therefore, heap dumps generated at runtime will contain content from both the "image heap" and the read-write heap.
```

gsmet commented on _posts/2024-04-15-heapdump-oome-native.adoc:
```diff
@@ -0,0 +1,265 @@
+
+Multiple graphical tools exist out there to inspect heap dumps,
+such as https://visualvm.github.io/[VisualVM],
+https://eclipse.dev/mat/[Eclipse Memory Analyzer (MAT)]...etc,
```
```suggestion
https://eclipse.dev/mat/[Eclipse Memory Analyzer (MAT)], etc.,
```

gsmet commented on _posts/2024-04-15-heapdump-oome-native.adoc:
```diff
@@ -0,0 +1,265 @@
+---
+layout: post
+title: 'Obtaining heap dump on OutOfMemoryError with Quarkus native'
+date: 2024-04-15
```
Let's target a publication on Tuesday? We can merge tomorrow and get it published automatically on Tuesday and tweet about it on Tuesday? Friday is a slow day for blog posts.

Note that if you want it published tomorrow, I'm around!

galderz replied:
Tuesday is fine, I'll ammend it.

gsmet reviewed (commented):
Added some minor comments. Let me know when you prefer it to be published, I'll make it happen.

gsmet reviewed (approved):
Awesome, thanks!

gsmet commented:
If you write me a tweet, I will be happy to post it on Tuesday.

galderz commented:
> If you write me a tweet, I will be happy to post it on Tuesday.

Sure, how about this?

> Explore heap dumps and learn how to use HeapDumpOOME option in Quarkus native apps.

gsmet commented:
@galderz What's your Twitter handle?

galderz commented:
> @galderz What's your Twitter handle?

@galderz :)
